### PR TITLE
feat(manager): add mise package manager

### PR DESCRIPTION
**Pull request by james0209**
## Changes

Add support for the `mise` package manager: https://mise.jdx.dev/getting-started.html

`mise` has it's own TOML file - `.mise.toml` - which is an extension of `asdf`'s `.tool-versions`

## Context

Closes https://github.com/renovatebot/renovate/issues/28918

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/mise/__fixtures__/Mise.1.toml
Comment (rarkins) [inline lib/modules/manager/mise/__fixtures__/Mise.1.toml:1]: Please inline these small fixtures using `codeBlock`

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/mise/extract.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.ts:18]: Use a zod schema instead of `as`

File: lib/modules/manager/mise/extract.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.ts:57]: What does it mean, if multiple tool versions are defined here?

File: lib/modules/manager/mise/extract.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.ts:60]: use `null`, if manually set.

File: lib/modules/manager/mise/extract.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.ts:66]:
```suggestion
): ToolingConfig | null {
```
Here too.

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/mise/extract.ts
Comment (james0209) [reply lib/modules/manager/mise/extract.ts:51]: I believe it follows the same logic as `asdf`'s `.tool-versions` ([ref](https://asdf-vm.com/manage/configuration.html#tool-versions))

> Multiple versions can be set by separating them with a space. For example, to use Python 3.7.2, fallback to Python 2.7.15 and finally to the system Python, the following line can be added to .tool-versions.
> 
> 
> `python 3.7.2 2.7.15 system`

Example in mise documentation [here](https://mise.jdx.dev/configuration.html#mise-toml)

```
[tools]
# specify single or multiple versions
terraform = '1.0.0'
erlang = ['23.3', '24.0']

# supports everything you can do with .tool-versions currently
node = ['16', 'prefix:20', 'ref:master', 'path:~/.nodes/14']
```

The logic used for `asdf` is to only update the first version if multiple are listed, as the rest are fallbacks ([ref](https://github.com/renovatebot/renovate/blob/f19afb4dee5e5bf5b7ef6a67d6d817f2d38f6773/lib/modules/manager/asdf/readme.md?plain=1#L11)) - so I've used the same logic in this PR.

File: lib/modules/manager/mise/index.ts
Comment (viceice) [inline lib/modules/manager/mise/index.ts:1]:
```suggestion
import { supportedDatasources as asdfSupportedDatasources } from '../asdf';
```

File: lib/modules/manager/mise/upgradeable-tooling.ts
Comment (viceice) [inline lib/modules/manager/mise/upgradeable-tooling.ts:53]: can we reuse the logic from asdf?

File: lib/modules/manager/mise/utils.ts
Comment (viceice) [inline lib/modules/manager/mise/utils.ts:9]: pass the packafilename and log it here

File: lib/modules/manager/mise/upgradeable-tooling.ts
Comment (james0209) [reply lib/modules/manager/mise/upgradeable-tooling.ts:53]: Theoretically yes, but there's some issues that I think makes it not worthwhile.

The `asdf` Java plugin does not support all the same types of Java as the `mise` one. i.e. in the `asdf` manager logic, there is no pathways for Zulu Java (`zulu-`), Corretto Java (`corretto-`), or even just a version number (i.e. `java = "17.0.2"`)

- This becomes an issue as `mise` supports having no prefix - defaulting to OpenJDK for no prefix.
- There are plenty of example of these different types of Java / no prefix being used in `.mise.toml` files. You can see on Github Search [here](https://github.com/search?q=mise.toml+java+language%3ATOML&type=code&l=TOML).

File: lib/modules/manager/mise/extract.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.ts:56]:
```suggestion
  if (is.nonEmptyString(toolData)) {
    // Handle the string case
    // e.g. 'erlang = "23.3"'
    return toolData;
  }
    if (is.array(toolData, is.string)) {
    // Handle the array case
    // e.g. 'erlang = ["23.3", "24.0"]'
    return toolData[0]; // Get the first version in the array
  }
  if (is.nonEmptyString(toolData.version)) {
    // Handle the object case with a string version
    // e.g. 'python = { version = "3.11.2" }'
    return toolData.version;
  } 
```
Simplify

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/mise/extract.ts
Comment (james0209) [reply lib/modules/manager/mise/extract.ts]: Nice suggestion, thanks!

File: lib/modules/manager/mise/extract.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.ts:57]:
```suggestion
  if (is.nonEmptyString(toolData.version)) {
    // Handle the object case with a string version
    // e.g. 'python = { version = "3.11.2" }'
    return toolData.version;
  }
```
Because of type narrowing we already know that it is a object and based on your implementation this should return null for empty string

File: lib/modules/manager/mise/extract.spec.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.spec.ts:174]:
```suggestion
      erlang = []
```
Add a test that empty arrays are handled gracefully.

File: lib/modules/manager/mise/extract.ts
Comment (secustor) [inline lib/modules/manager/mise/extract.ts:49]:
```suggestion
    return toolData.length ? toolData[0] : null; // Get the first version in the array
```

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/mise/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mise/readme.md:13]:
```suggestion
Renovate's `mise` manager can version these tools:

<!-- Autogenerate in https://github.com/renovatebot/renovate -->
<!-- Autogenerate end -->

<!-- prettier-ignore -->
!!! warning
    Only the _first_ version entry for each supported tool is managed!
    This is due to `.mise.toml` supporting fallback versions.

To understand the versioning, Renovate re-uses:

- [mise's plugins](https://github.com/jdx/mise/tree/main/src/plugins/core)
- and [asdf's plugins](https://mise.jdx.dev/registry.html)

This also means that support for a new tool's versioning must be added one-by-one.
```

Changes:

- Put the important info for the reader up front, so the list of supported tools goes on top
- Bump the alert level of the admonition to a `warning` to grab attention better
- Make a list of the plugins Renovate re-uses
- Write short sentences
- Use simple words and phrases
- Split long sentences into two

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/manager/mise/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mise/readme.md:11]: It feels we're missing a step in the explanation. Here's what I think when I read the current message:

1. Okay, so only the _first_ version for each tool is managed.
2. Looks like `.mise.toml` fallback versions cause this behavior.
3. But "why/how" does the fallback version "break" the way Renovate "sees/upgrades" versions?

Can you please improve the explanation, so the logical leap from "fallback version" to "broken Renovate" is clear? :wink:

File: lib/modules/manager/mise/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mise/readme.md:14]: Thank you for the extra explanation! :heart: The text is now too big for a admonition. So let's give the text some space, and nice headings. I also attempted another rewrite, to use simpler words and phrases. Feel free to correct any technical errors. :wink:

Do you think we should recommend users only use a _single_ version? Or will that cause more problems?

Here's my suggested text, for you to copy/paste/edit:

```markdown
### Renovate only updates primary versions

Renovate only updates the _first_ version it finds for each tool listed in the `mise.toml` file.
You must manually update any fallback versions!

For example if you had this snippet in your `mise.toml` file:

`'erlang = ["23.3", "22.0"]'`

Then Renovate would update the first, primary version: `"23.3"` but ignore the fallback `"22.0"` version.

#### Why can Renovate only update primary versions?

Usually the mise package manager can access the primary version, and skips any fallback versions.

But when the mise package manager accesses a fallback version, Renovate gets confused.

To be safe, Renovate only updates the _first_ version it sees.

## Renovate re-uses these plugins
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by secustor — DISMISSED** [review]
Waiting for a review `@HonkingGoose` before merging.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
The docs are nearly good to go, some small style fixes.

Changes:

- Swap some difficult words for simple words
- Drop `.` at the end of list items
- Rewrite some sentences
- Use `monospace` styling for some things

By the way, we have a [Style Guide](https://github.com/renovatebot/renovate/blob/main/docs/development/style-guide.md) for user-facing text. That explains why I'm making some changes, that may look weird at first, like the `.` at the end of lists. :wink:

File: lib/modules/manager/mise/readme.md
Comment (james0209) [reply lib/modules/manager/mise/readme.md]: Rewritten it a lil' bit - mise uses asdf's plugin ecosystem under the hood, so adding support for `asdf-x` plugins is arguably one of the easiest ways to add support for tooling into `mise`.

`mise` does have core plugins which "are new and will improve over time. ([ref](https://mise.jdx.dev/plugins.html#core-plugins)), but I'm not sure about user contribution to them (i.e. it looks like it's only for main language versions rather than tools generically)

File: lib/modules/manager/mise/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mise/readme.md:33]:
```suggestion
### Plugin/tool support
```

File: lib/modules/manager/mise/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mise/readme.md:51]:
```suggestion
If `mise` adds support for more tools via its own [core plugins](https://mise.jdx.dev/plugins.html#core-plugins), you can create a PR to extend Renovate's `mise` manager to add support for the new tooling.
```

File: lib/modules/manager/mise/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mise/readme.md:53]:
```suggestion
You may be able to add support for new tooling upstream in the core plugins - create an issue and see if the community agrees whether it belongs there, or if it would be better as an `asdf-` plugin.
```

Let's drop the bullet point here, and keep it as plain text.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Final review, I promise! :wink:

File: lib/modules/manager/mise/readme.md
Comment (HonkingGoose) [reply lib/modules/manager/mise/readme.md]: I'm happy with your text. I don't know enough to decide which method is best, so I'll let you judge what's best. :smile: 

Readers can now use the information in the docs to decide how they want to add support for new tools to Renovate.

**Review by HonkingGoose — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Comment by james0209** [comment]
@rarkins can you give this PR some 👀 please? It can't be merged until you re-review it - thanks

**Review by rarkins — APPROVED** [review]

**Pull request merged**
